windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+            long binaryResourcesModifiedTime = await GetBinaryResourcesModifiedTime();
+            StorageFile packageFile = await codePushPackage.GetCurrentPackageBundle(this.assetsBundleFileName);
+            if (packageFile == null)
+            {
+                // There has not been any downloaded updates.
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+
+            JObject packageMetadata = await codePushPackage.GetCurrentPackage();
+            long? binaryModifiedDateDuringPackageInstall = null;
+            string binaryModifiedDateDuringPackageInstallString = (string)packageMetadata[BINARY_MODIFIED_TIME_KEY];
+            if (binaryModifiedDateDuringPackageInstallString != null)
+            {
+                binaryModifiedDateDuringPackageInstall = long.Parse(binaryModifiedDateDuringPackageInstallString);
+            }
+
+            string packageAppVersion = (string)packageMetadata["appVersion"];
+
+            // TODO: test configuration
+            if (binaryModifiedDateDuringPackageInstall != null &&
+                    binaryModifiedDateDuringPackageInstall == binaryResourcesModifiedTime &&
+                    (IsUsingTestConfiguration() || appVersion.Equals(packageAppVersion)))
+            {
+                CodePushUtils.logBundleUrl(packageFile.Path);
+                isRunningBinaryVersion = false;
+                return FILE_BUNDLE_PREFIX + packageFile.Path.Replace(ApplicationData.Current.LocalFolder.Path, "").Replace("\\", "/");
+            }
+            else
+            {
+                // The binary version is newer.
+                didUpdate = false;
+                if (!mainPage.UseDeveloperSupport || !appVersion.Equals(packageAppVersion))
+                {
+                    await ClearUpdates();
+                }
+
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+        }
+
+        private ApplicationDataContainer GetCodePushSettings()
+        {
+            return ApplicationData.Current.LocalSettings.CreateContainer(CODE_PUSH_PREFERENCES, ApplicationDataCreateDisposition.Always);
+        }
+
+        private JArray GetFailedUpdates()
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            string failedUpdatesString = (string)settings.Values[FAILED_UPDATES_KEY];
+            if (failedUpdatesString == null)
+            {
+                return new JArray();
+            }
+
+            try
+            {
+                return JArray.Parse(failedUpdatesString);
+            }
+            catch (Exception)
+            {
+                JArray emptyArray = new JArray();
+                settings.Values[FAILED_UPDATES_KEY] = JsonConvert.SerializeObject(emptyArray);
```

Comment:
my understanding of it is "if we can't parse the failedUpdates from the settings, then replace the failedUpdates in the settings with an empty JSON object". makes sense to me if we're sure that the only reason JArray.parse could fail is that failedUpdatesString is filled with invalid data, and that we're not overwriting good data that just failed to parse once due to some other issue